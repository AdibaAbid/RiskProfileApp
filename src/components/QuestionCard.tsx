import React from 'react';
import {View, Text, TouchableOpacity, StyleSheet} from 'react-native';
import theme from '../theme';

interface Question {
  id: number;
  questionText: string;
  options: string[];
  points?: number[];
}

interface QuestionCardProps {
  question: Question;
  questionIndex: number;
  selectedOption: number | null | undefined;
  onOptionSelect: (questionId: number, optionIndex: number) => void;
}

const QuestionCard: React.FC<QuestionCardProps> = ({
  question,
  questionIndex,
  selectedOption,
  onOptionSelect,
}) => {
  return (
    <View style={styles.questionBlock}>
      <Text style={styles.questionText}>
        {questionIndex + 1}. {question.questionText}
      </Text>
      {question.options.map((option, index) => (
        <TouchableOpacity
          key={index}
          style={[
            styles.optionButton,
            selectedOption === index && styles.selectedOption,
          ]}
          testID={`option-${question.id}-${index}`}
          onPress={() => onOptionSelect(question.id, index)}>
          <Text
            style={[
              styles.optionText,
              selectedOption === index && styles.selectedOptionText,
            ]}>
            {option}
          </Text>
        </TouchableOpacity>
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  questionBlock: {
    marginBottom: 25,
    backgroundColor: theme.white,
    padding: 15,
    borderRadius: 10,
    shadowColor: theme.black,
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 0.1,
    shadowRadius: 4,
    elevation: 3,
  },
  questionText: {
    fontSize: 18,
    fontWeight: '600',
    marginBottom: 12,
    color: theme.darkestBlack,
  },
  optionButton: {
    paddingVertical: 12,
    paddingHorizontal: 15,
    backgroundColor: theme.lightestGrey,
    marginVertical: 5,
    borderRadius: 8,
    borderWidth: 1,
    borderColor: theme.subtleGrey,
  },
  selectedOption: {
    backgroundColor: theme.pleasentBlue,
    borderColor: theme.paleBlue,
  },
  optionText: {
    fontSize: 16,
    color: theme.darkGrey,
  },
  selectedOptionText: {
    color: theme.white,
    fontWeight: '500',
  },
});

export default QuestionCard;
