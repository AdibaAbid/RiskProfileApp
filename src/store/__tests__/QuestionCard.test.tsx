import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import QuestionCard from '../../components/QuestionCard';

jest.mock('../../theme', () => ({
  white: '#ffffff',
  black: '#000000',
  darkestBlack: '#333333',
  lightestGrey: '#e9ecef',
  subtleGrey: '#ced4da',
  pleasentBlue: '#007bff',
  paleBlue: '#0056b3',
  darkGrey: '#212529',
}));

const mockQuestion = {
  id: 1,
  questionText: 'What is your favorite color?',
  options: ['Red', 'Green', 'Blue'],
};

const mockOnOptionSelect = jest.fn();

describe('QuestionCard', () => {
  beforeEach(() => {
    mockOnOptionSelect.mockClear();
  });

  it('renders the question text and options correctly', () => {
    const {getByText} = render(
      <QuestionCard
        question={mockQuestion}
        questionIndex={0}
        selectedOption={null}
        onOptionSelect={mockOnOptionSelect}
      />,
    );

    expect(getByText('1. What is your favorite color?')).toBeTruthy();
    expect(getByText('Red')).toBeTruthy();
    expect(getByText('Green')).toBeTruthy();
    expect(getByText('Blue')).toBeTruthy();
  });

  it('calls onOptionSelect with correct arguments when an option is pressed', () => {
    const {getByText} = render(
      <QuestionCard
        question={mockQuestion}
        questionIndex={0}
        selectedOption={null}
        onOptionSelect={mockOnOptionSelect}
      />,
    );

    const redOption = getByText('Red');
    fireEvent.press(redOption);

    expect(mockOnOptionSelect).toHaveBeenCalledTimes(1);
    expect(mockOnOptionSelect).toHaveBeenCalledWith(mockQuestion.id, 0); // 0 is the index of 'Red'

    const blueOption = getByText('Blue');
    fireEvent.press(blueOption);
    expect(mockOnOptionSelect).toHaveBeenCalledTimes(2);
    expect(mockOnOptionSelect).toHaveBeenCalledWith(mockQuestion.id, 2); // 2 is the index of 'Blue'
  });

  it('applies selected styles to the selected option', () => {
    const {getByText, rerender} = render(
      <QuestionCard
        question={mockQuestion}
        questionIndex={0}
        selectedOption={null} // Initially no option selected
        onOptionSelect={mockOnOptionSelect}
      />,
    );

    expect(getByText('Green')).toBeTruthy();

    rerender(
      <QuestionCard
        question={mockQuestion}
        questionIndex={0}
        selectedOption={1} // 'Green' is selected
        onOptionSelect={mockOnOptionSelect}
      />,
    );

    const greenOption = getByText('Green');
    expect(greenOption).toBeTruthy();
  });

  it('renders correctly when a selectedOption prop is provided', () => {
    const {getByText} = render(
      <QuestionCard
        question={mockQuestion}
        questionIndex={0}
        selectedOption={2} // 'Blue' is selected
        onOptionSelect={mockOnOptionSelect}
      />,
    );
    expect(getByText('Blue')).toBeTruthy();
  });

  it('matches snapshot when no option is selected', () => {
    const tree = render(
      <QuestionCard
        question={mockQuestion}
        questionIndex={0}
        selectedOption={null}
        onOptionSelect={mockOnOptionSelect}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when an option is selected', () => {
    const tree = render(
      <QuestionCard
        question={mockQuestion}
        questionIndex={0}
        selectedOption={1}
        onOptionSelect={mockOnOptionSelect}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
