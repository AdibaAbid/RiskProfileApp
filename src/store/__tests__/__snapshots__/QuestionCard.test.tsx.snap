// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionCard matches snapshot when an option is selected 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "borderRadius": 10,
      "elevation": 3,
      "marginBottom": 25,
      "padding": 15,
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 4,
    }
  }
>
  <Text
    style={
      {
        "color": "#333333",
        "fontSize": 18,
        "fontWeight": "600",
        "marginBottom": 12,
      }
    }
  >
    1
    . 
    What is your favorite color?
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#e9ecef",
        "borderColor": "#ced4da",
        "borderRadius": 8,
        "borderWidth": 1,
        "marginVertical": 5,
        "opacity": 1,
        "paddingHorizontal": 15,
        "paddingVertical": 12,
      }
    }
    testID="option-1-0"
  >
    <Text
      style={
        [
          {
            "color": "#212529",
            "fontSize": 16,
          },
          false,
        ]
      }
    >
      Red
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#007bff",
        "borderColor": "#0056b3",
        "borderRadius": 8,
        "borderWidth": 1,
        "marginVertical": 5,
        "opacity": 1,
        "paddingHorizontal": 15,
        "paddingVertical": 12,
      }
    }
    testID="option-1-1"
  >
    <Text
      style={
        [
          {
            "color": "#212529",
            "fontSize": 16,
          },
          {
            "color": "#ffffff",
            "fontWeight": "500",
          },
        ]
      }
    >
      Green
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#e9ecef",
        "borderColor": "#ced4da",
        "borderRadius": 8,
        "borderWidth": 1,
        "marginVertical": 5,
        "opacity": 1,
        "paddingHorizontal": 15,
        "paddingVertical": 12,
      }
    }
    testID="option-1-2"
  >
    <Text
      style={
        [
          {
            "color": "#212529",
            "fontSize": 16,
          },
          false,
        ]
      }
    >
      Blue
    </Text>
  </View>
</View>
`;

exports[`QuestionCard matches snapshot when no option is selected 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "borderRadius": 10,
      "elevation": 3,
      "marginBottom": 25,
      "padding": 15,
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 4,
    }
  }
>
  <Text
    style={
      {
        "color": "#333333",
        "fontSize": 18,
        "fontWeight": "600",
        "marginBottom": 12,
      }
    }
  >
    1
    . 
    What is your favorite color?
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#e9ecef",
        "borderColor": "#ced4da",
        "borderRadius": 8,
        "borderWidth": 1,
        "marginVertical": 5,
        "opacity": 1,
        "paddingHorizontal": 15,
        "paddingVertical": 12,
      }
    }
    testID="option-1-0"
  >
    <Text
      style={
        [
          {
            "color": "#212529",
            "fontSize": 16,
          },
          false,
        ]
      }
    >
      Red
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#e9ecef",
        "borderColor": "#ced4da",
        "borderRadius": 8,
        "borderWidth": 1,
        "marginVertical": 5,
        "opacity": 1,
        "paddingHorizontal": 15,
        "paddingVertical": 12,
      }
    }
    testID="option-1-1"
  >
    <Text
      style={
        [
          {
            "color": "#212529",
            "fontSize": 16,
          },
          false,
        ]
      }
    >
      Green
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#e9ecef",
        "borderColor": "#ced4da",
        "borderRadius": 8,
        "borderWidth": 1,
        "marginVertical": 5,
        "opacity": 1,
        "paddingHorizontal": 15,
        "paddingVertical": 12,
      }
    }
    testID="option-1-2"
  >
    <Text
      style={
        [
          {
            "color": "#212529",
            "fontSize": 16,
          },
          false,
        ]
      }
    >
      Blue
    </Text>
  </View>
</View>
`;
